bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/internal/DefaultSWTFontRegistry.java

```diff
@@ -72,7 +72,8 @@ public Font getFont(FontData fontData, int zoom) {
 	}
 
 	private Font registerFont(FontData fontData, Font font) {
-		fontsMap.put(fontData, font);
+		FontData clonedFontData = new FontData(fontData.toString());
```

Comment:
This looks like a heavy-weight operation, creating a String just to parse it again.
Isn't there a (internal) copy-constructor or copy method that handles this more efficiently?

Reply:
I agree that it's inefficient but there is no internal constructor to do it. 

In my poor man's performance test, cloning fonts and adding them to the registry starts slow (700 ms) and becomes faster with time, averaging 106 ms per call in the 10th call and 65 ms in the 20th.

Reply:
Can't we make `FontData` implement `Clonable` and then make this here 
```suggestion
		FontData clonedFontData = fontData.clone();
```
as an alternative we could have a copy construction in wich case it becomes

```suggestion
		FontData clonedFontData = new FontData(fontData);
```

Even if that would result in clone being implemented that way (calling toString()), it would be more semantically and allows to maybe being implemented for efficient in SWT.

Reply:
It would be possible, but really: we are talking about an insignificant amount of time at the moment and the values do get better (faster) every time one instantiates a `FontData`. I just added a few debug lines to the constructor of `FontData` and ran it in my IDE and this is what I got:

<details><summary>Avg. instantiation times of FontData</summary>
<p>

```bash
CompileCommand: exclude org/eclipse/jdt/internal/core/dom/rewrite/ASTRewriteAnalyzer.getExtendedRange bool exclude = true
WARNING: Using incubator modules: jdk.incubator.vector
Apr. 14, 2025 12:02:59 PM org.apache.aries.spifly.BaseActivator log
INFO: Registered provider org.slf4j.simple.SimpleServiceProvider of service org.slf4j.spi.SLF4JServiceProvider in bundle slf4j.simple
1: avg=95000 nanoseconds
2: avg=58400 nanoseconds
WARNING: Annotation classes from the 'javax.inject' or 'javax.annotation' package found.
It is recommended to migrate to the corresponding replacements in the jakarta namespace.
The Eclipse E4 Platform will remove support for those javax-annotations in a future release.
To suppress this warning, set the VM property: -Declipse.e4.inject.javax.warning=false
To disable processing of 'javax' annotations entirely, set the VM property: -Declipse.e4.inject.javax.disabled=true

3: avg=44066 nanoseconds
4: avg=35975 nanoseconds
5: avg=30900 nanoseconds
6: avg=27316 nanoseconds
7: avg=24657 nanoseconds
8: avg=22687 nanoseconds
9: avg=24855 nanoseconds
10: avg=23220 nanoseconds
11: avg=24645 nanoseconds
12: avg=25275 nanoseconds
13: avg=24215 nanoseconds
14: avg=23150 nanoseconds
15: avg=22206 nanoseconds
16: avg=22843 nanoseconds
17: avg=21929 nanoseconds
18: avg=21211 nanoseconds
19: avg=20484 nanoseconds
20: avg=19820 nanoseconds
21: avg=19219 nanoseconds
22: avg=18677 nanoseconds
23: avg=18178 nanoseconds
24: avg=17712 nanoseconds
25: avg=17292 nanoseconds
26: avg=16896 nanoseconds
27: avg=16540 nanoseconds
28: avg=16203 nanoseconds
29: avg=15893 nanoseconds
30: avg=15596 nanoseconds
31: avg=15325 nanoseconds
32: avg=16284 nanoseconds
33: avg=16575 nanoseconds
34: avg=17411 nanoseconds
35: avg=17728 nanoseconds
Warning: NLS missing message: MarketplacePage_favorites in: org.eclipse.epp.internal.mpc.ui.wizards.messages
SLF4J(W): No SLF4J providers were found.
SLF4J(W): Defaulting to no-operation (NOP) logger implementation
SLF4J(W): See https://www.slf4j.org/codes.html#noProviders for further details.
36: avg=18530 nanoseconds
37: avg=18391 nanoseconds
38: avg=18368 nanoseconds
39: avg=18100 nanoseconds
40: avg=17885 nanoseconds
41: avg=17592 nanoseconds
42: avg=17302 nanoseconds
43: avg=17174 nanoseconds
44: avg=16911 nanoseconds
45: avg=16755 nanoseconds
46: avg=16619 nanoseconds
47: avg=16476 nanoseconds
48: avg=16387 nanoseconds
49: avg=16202 nanoseconds
50: avg=15994 nanoseconds
51: avg=15813 nanoseconds
52: avg=15621 nanoseconds
53: avg=15447 nanoseconds
54: avg=15424 nanoseconds
55: avg=15320 nanoseconds
56: avg=15180 nanoseconds
57: avg=15031 nanoseconds
58: avg=15179 nanoseconds
59: avg=15013 nanoseconds
60: avg=14846 nanoseconds
61: avg=14870 nanoseconds
62: avg=14820 nanoseconds
63: avg=14747 nanoseconds
64: avg=14603 nanoseconds
65: avg=14538 nanoseconds
66: avg=14422 nanoseconds
67: avg=14362 nanoseconds
68: avg=14227 nanoseconds
69: avg=14097 nanoseconds
70: avg=14011 nanoseconds
71: avg=14195 nanoseconds
72: avg=14190 nanoseconds
73: avg=14117 nanoseconds
74: avg=14018 nanoseconds
75: avg=13917 nanoseconds
76: avg=13838 nanoseconds
77: avg=13746 nanoseconds
78: avg=13675 nanoseconds
79: avg=13563 nanoseconds
80: avg=13551 nanoseconds
81: avg=13514 nanoseconds
82: avg=13682 nanoseconds
83: avg=13640 nanoseconds
84: avg=13565 nanoseconds
85: avg=13496 nanoseconds
86: avg=13425 nanoseconds
87: avg=13334 nanoseconds
88: avg=13271 nanoseconds
89: avg=13219 nanoseconds
90: avg=13177 nanoseconds
91: avg=13315 nanoseconds
92: avg=13255 nanoseconds
93: avg=13201 nanoseconds
94: avg=13128 nanoseconds
95: avg=13118 nanoseconds
96: avg=13048 nanoseconds
97: avg=13061 nanoseconds
98: avg=13034 nanoseconds
99: avg=12981 nanoseconds
100: avg=12913 nanoseconds
101: avg=12843 nanoseconds
102: avg=12774 nanoseconds
104: avg=13060 nanoseconds
105: avg=13038 nanoseconds
106: avg=13016 nanoseconds
107: avg=13017 nanoseconds
108: avg=12988 nanoseconds
109: avg=12933 nanoseconds
110: avg=12864 nanoseconds
111: avg=12838 nanoseconds
112: avg=12794 nanoseconds
113: avg=12730 nanoseconds
114: avg=12747 nanoseconds
115: avg=12723 nanoseconds
116: avg=12662 nanoseconds
117: avg=12668 nanoseconds
118: avg=12627 nanoseconds
119: avg=12594 nanoseconds
120: avg=12558 nanoseconds
121: avg=12520 nanoseconds
122: avg=12475 nanoseconds
123: avg=12504 nanoseconds
124: avg=12515 nanoseconds
125: avg=12514 nanoseconds
126: avg=12456 nanoseconds
127: avg=12422 nanoseconds
128: avg=12389 nanoseconds
129: avg=12369 nanoseconds
130: avg=12332 nanoseconds
131: avg=12270 nanoseconds
132: avg=12263 nanoseconds
133: avg=12266 nanoseconds
134: avg=12406 nanoseconds
135: avg=12410 nanoseconds
136: avg=12386 nanoseconds
137: avg=12345 nanoseconds
138: avg=12294 nanoseconds
139: avg=12248 nanoseconds
140: avg=12199 nanoseconds
141: avg=12213 nanoseconds
142: avg=12197 nanoseconds
143: avg=12162 nanoseconds
144: avg=12128 nanoseconds
145: avg=12090 nanoseconds
146: avg=12076 nanoseconds
147: avg=12039 nanoseconds
148: avg=11993 nanoseconds
149: avg=11979 nanoseconds
150: avg=11942 nanoseconds
151: avg=11919 nanoseconds
152: avg=11880 nanoseconds
153: avg=11840 nanoseconds
154: avg=11798 nanoseconds
155: avg=11756 nanoseconds
156: avg=11712 nanoseconds
157: avg=11699 nanoseconds
158: avg=11669 nanoseconds
159: avg=11633 nanoseconds
160: avg=11620 nanoseconds
161: avg=11621 nanoseconds
162: avg=11622 nanoseconds
163: avg=11604 nanoseconds
164: avg=11592 nanoseconds
165: avg=11575 nanoseconds
166: avg=11550 nanoseconds
167: avg=11527 nanoseconds
168: avg=11496 nanoseconds
169: avg=11455 nanoseconds
170: avg=11419 nanoseconds
171: avg=11383 nanoseconds
172: avg=11345 nanoseconds
173: avg=11332 nanoseconds
174: avg=11324 nanoseconds
```
</p>
</details> 

And here's what I did to test it:
- Started my IDE
- Made the font in the Java editor bigger/smaller a couple of times with `Ctrl + +`  /  `Ctrl + -`
- Changed the font in the preferences
- Increased/Decreased the font again in the editor

There was absolutely no perceptible performance degradation and the console output also shows that the worst-case scenario (the 1st instantiation) took **95000 nanoseconds (< 1 ms)**.

Reply:
I basically agree with @laeubi here. I had the same comment about using the String-based constructor but unfortunately forgot to submit my review before going on vacation. Constructing a font data object based on the `toString()` representation of another is not that good from design perspective. But it is guaranteed to work by contract of the constructor and the `toString()` implementation. There is even a test for this, as the `toString()` representation seems to be used for data transfer at other places as well. Given that I would be fine with just keeping everything as is (also considering that performance is not an issue here), even though the design is not that great.

Reply:
@laeubi I understand and agree with most of it, but it's been like this since 2001 (https://github.com/eclipse-platform/eclipse.platform.swt/commit/3275970ab8046022b8d2069b5c21bf117c223db7#diff-53bdc71379e2933db1c054bf121212ab1449d1e19dbc9a9cb56dd8ceb9f8979f) so it's a well established way of doing it for the consumers of SWT.

Reply:
Its never to late to improve :-)

Reply:
I'm sorry, can you take over and send me a PR? I'm working on other stuff for Edge right now.